Improved charts, Added provenance tab

SanjeevLakhwani opened the pull request:
## Improved data display for "Missing" field in bar charts/ histograms
- Coming from the issue that in a lot of data, missing field dominated the charts requiring large y-axis scales and harder to understand data, due to this now missing field would not be shown as a column but as a descriptive text with the chart card
- For more clarity it contains an person logo attached to the next but for future we would need to differentiate between individuals or experiments

## Added "Other" Section to pie charts 
- In really diversely distributed data, there existed really small sections which are hard to focus on, therefore created an Other section which combines all sections below a specific sector size, given that there are more than one such sectors

## Missing Sector in Pie chart now has a common color scheme
- Every Sector with key "Missing" (case insensitive) would now have a grey color

## Added a Provenance tab
- This Tab is to display information from where the data comes from
- It displays information given by katsu, which is the provenance for all the dataset held in katsu
- It resolves #26 
- It relies on `/api/public_dataset` for response. if the response contains a dats_file it renders that otherwise the parent object

## Added a max limit to number of charts displayed on first render
- Now you can limit the number of charts that would be displayed
  - Note: Local storage still overwrites whether a chart is displayed or not, so it is only valid when local storage isn't present
- To test with less number of charts, the default max being 9, change the variable `MAX_CHARTS` in configConstants.js to desired number and don't forget to clear localstorage!

ppillot reviewed (commented):
I've made a cursory look through the changes. I've noticed that there is a difference with bento_web in the way the "other" category feature is implemented:
- in bento_web the special case where **only one category** is bellow the threshold is detected so that no "Other" is displayed
- in this commit I don't think that the code makes this distinction

SanjeevLakhwani commented:
> I've made a cursory look through the changes. I've noticed that there is a difference with bento_web in the way the "other" category feature is implemented:
> 
> * in bento_web the special case where **only one category** is bellow the threshold is detected so that no "Other" is displayed
> * in this commit I don't think that the code makes this distinction

Thanks for pointing it out, I have made the change and also added another change for missing sector in pie charts. I have added details in the PR statement

ppillot commented on src/js/utils/string.js:
```diff
@@ -0,0 +1,2 @@
+export const toTitleCase = (str) =>
+  str.replace(/\w\S*/g, (txt) => txt.charAt(0).toUpperCase() + txt.substring(1).toLowerCase());
```
Not sure it is used again, but if you ever need to do this, consider the CSS alternative `text-transform: capitalize`

SanjeevLakhwani replied:
It is not used, therefore removed it. It was previously used for the older version of provenance but my bad missed it in the cleanup.

SanjeevLakhwani commented on src/js/components/overview/MakeChartCard.js:
```diff
@@ -29,6 +30,15 @@ const MakeChartCard = ({ section, chart, onRemoveChart }) => {
+    </Typography.Text>
+  );
+
+  if (missingCount) ed.unshift(missingDataText);
```
done

SanjeevLakhwani commented on src/js/components/overview/charts/BentoPie.js:
```diff
@@ -36,7 +36,18 @@ const labelShortName = (name) => {
 
-  data = data.filter((e) => e.y !== 0);
+  // Changing immutable data to mutable data
+  data = [...data];
```
Merci

SanjeevLakhwani commented on src/js/components/overview/charts/BentoPie.js:
```diff
@@ -36,7 +36,18 @@ const labelShortName = (name) => {
+  data = temp.length === length - 1 ? data : temp;
+  if (data.length !== length) {
+    data.push({ x: 'Other', y: sum - data.reduce((acc, e) => acc + e.y, 0) });
+  }
```
done

SanjeevLakhwani commented on src/js/components/Util/LinkIfUrl.js:
```diff
@@ -0,0 +1,16 @@
+
+// Renders text as link if the text provided is a valid url.
+const LinkIfUrl = ({ text }) => {
+  if (text.match(/^(http|https):\/\/[^ "]+$/)) {
```
done

SanjeevLakhwani commented on src/js/components/Provenance/Tables/IsAboutTable.js:
```diff
@@ -0,0 +1,30 @@
+        { title: 'Name', dataIndex: 'name', key: 'name' },
+        {
+          title: 'Identifier',
+          dataIndex: 'identifier.identifier',
```
left it as that for the reader to get the path but antd cannot access nested properties

SanjeevLakhwani commented on src/js/components/overview/charts/BentoPie.js:
```diff
@@ -36,7 +36,18 @@ const labelShortName = (name) => {
+  }
+
   const bentoFormatData = data.map((e) => ({ name: e.x, value: e.y }));
   const totalCount = bentoFormatData.reduce((sum, e) => sum + e.value, 0);
```
It is I changed it

ppillot commented:
Thanks for all these changes!
The API error reporting function is a nice addition too.
I've added some minor modification to handle the shape of the response coming from Katsu WRT datasets properties, and did some corrections to some comments.
Tested and approved!
